Node: replace tabix child processes with tabix Javascript wrapper

hannahle opened the pull request.

hannahle committed:
Added JS wrapper for tabix and fixed tabix GChr37 version

hannahle committed:
replaced tabix child process with a working tabix js wrapper

hannahle committed:
removed unused imports

hannahle committed:
added more columns

kevinlul commented on server/package.json:
```diff
@@ -59,6 +62,7 @@
     "mongodb-memory-server": "^7.4.3",
     "mongoose": "^6.0.8",
     "node-cache": "^5.1.2",
+    "node-fetch": "2",
```
Is this to avoid the ESM-only node-fetch 3 upgrade?

hannahle replied:
New commit should revert it back to ESM only!

kevinlul commented on server/package.json:
```diff
@@ -70,5 +74,6 @@
     }
-  }
+  },
+  "type": "module"
```
This changes this project from CJS to ESM. Is that supposed be part of the scope of this change?

hannahle replied:
Nope - this was some experimental code that should be removed. I pushed up a new commit that reverted it back to CJS.

kevinlul commented on server/src/resolvers/getVariantsResolver/utils/fetchCaddAnnotations.ts:
```diff
@@ -1,86 +1,91 @@
-  protPos: 'aaPos',
-};
+  const tbiIndexed = new TabixIndexedFile({
+    filehandle: new RemoteFile(annotationUrl, { fetch: fetch as Fetcher }),
```
Are you redownloading this dependency every time Tabix is used?

hannahle replied:
The file is >300GB, but the query happens instantly, so no, I don't think the file specified with the `annotationUrl` is being downloaded every single time. I'm not entirely sure how tabix works under the hood, but according to the documentation [here](http://www.htslib.org/doc/tabix.html#:~:text=Last%20but%20not%20the%20least,without%20downloading%20the%20entire%20file.):

> Last but not the least, tabix supports remote data retrieval. One can put the data file and the index at an FTP or HTTP server, and other users or even web services will be able to get a slice without downloading the entire file.

kevinlul replied:
Okay, it's using HTTP byte-range headers to fetch chunks of it then.

hannahle committed:
Reverted ESM to CommonJS

kevinlul commented:
So the project is CJS, but an ESM-only node-fetch was imported?

hannahle committed:
Changed node-fetch version to 2 to allow for CommonJS import

hannahle committed:
Added types version

hannahle committed:
Changed from node-fetch to cross-fetch

hannahle commented:
> So the project is CJS, but an ESM-only node-fetch was imported?

Yes, this was the case. I also realized upon restarting the app that this is not allowed. I removed the node-fetch package altogether and replaced it with cross-fetch, which was already installed.

kevinlul commented on server/src/resolvers/getVariantsResolver/utils/fetchCaddAnnotations.ts:
```diff
@@ -1,86 +1,88 @@
+    1: 'pos',
+    2: 'ref',
+    3: 'alt',
+    7: 'consequence',
+    16: 'aaRef',
+    17: 'aaAlt',
+    19: 'transcript',
+    24: 'cdna',
+    28: 'aaPos',
+  };
 
-  return annotationsArray.map(annotation =>
-    annotation
-      .split(/\W+/)
-      .reduce<CaddAnnotation>(
-        (acc, curr, i) => ({ ...acc, [headers[i]]: curr }),
-        {} as CaddAnnotation
-      )
-  );
-};
+  const headersIndex = Object.keys(headersMap);
+  const headers = Object.values(headersMap);
 
-const _buildQuery = async (position: string, assemblyId: string) => {
-  const resolvedAssemblyId = resolveAssembly(assemblyId);
-  const annotationUrl = resolvedAssemblyId === '38' ? ANNOTATION_URL_38 : ANNOTATION_URL_37;
-  const indexPath = resolvedAssemblyId === '38' ? INDEX_38_PATH : INDEX_37_PATH;
+  const result = annotations.map(annotation => {
+    // Get only the required annotation columns
+    const annotationColumns = annotation
+      .split('\t')
+      .filter((a, i) => headersIndex.some(headerId => Number(headerId) === i));
+    return Object.fromEntries(
+      headers.map((h, i) => [h, annotationColumns[i]])
+    ) as unknown as CaddAnnotation;
+  });
 
-  return `tabix -h  ${annotationUrl} ${indexPath} ${position} | awk 'NR!=1{print $1,$2,$3,$4,$8,$17,$18,$20,$25,$29}'`;
+  console.log(result);
+  return result;
```
Can you explain what this function is supposed to do?

hannahle replied:
The tabix JS wrapper returns a list of strings, each string represents a line in the annotation tsv from `ANNOTATION_URL`, i.e. a variant annotation. It looks something like this: 

`19\t12312321\t23432\t....`

We know what headers correspond to what values based on CADD's documentation [here](https://cadd.gs.washington.edu/static/ReleaseNotes_CADD_v1.6.pdf).

The goal of the function is to take in a list of strings and return a list of objects of type `CaddAnnotation`.

hannahle replied:
I will add more detailed comments to the code.

kevinlul commented on server/src/resolvers/getVariantsResolver/utils/resolveChromosome.ts:
```diff
@@ -0,0 +1,14 @@
+const resolveChromosome = (position: string) => {
+  const chromosome = position.match(/^[^:]*[^ :]/gm)?.join('');
+  const coordinates = position.match(/\w[^:]*$/gm)?.join('');
+  const end = coordinates?.match(/\w[^-]*$/gm);
+  const start = coordinates?.match(/^[^-]*[^ -]/gm);
```
Were these expressions retrieved from elsewhere or did you write them yourself?

hannahle replied:
The last two were written by me once I understand what the individual components like /, ^, [], * do.

The first two I was able to get by consulting this Stackoverflow here: https://stackoverflow.com/questions/3858519/regular-expression-to-match-everything-that-appears-before-a-colon

(Hopefully one day I will be able to write so much regex that I can pull them out of my brain without looking at anything...)

kevinlul replied:
Yeah, this is my main concern with regex, which is that it becomes incomprehensible even to the original writer some time after it was written. Often, they aren't even the right tool for the job because of their complexity and runtime (and risks of catastrophic backtracking). Thus I would recommend against them if the objective is achievable with more understandable code; for example, everything before a colon can also be obtained from `.split(":")`.

hannahle replied:
That's a good point. I'll modify the code with split.

hannahle committed:
Added more detaild comments to CADD annotation function

hannahle committed:
Removed logs

hannahle committed:
simplified formatAnnotations function

hannahle committed:
Made a note about zero based versus 1 based index

kevinlul commented on server/src/utils/typedObject.ts:
```diff
@@ -0,0 +1,3 @@
+export function fromEntries<T>(entries: [keyof T, T[keyof T]][]): T {
+  return entries.reduce((acc, [key, value]) => ({ ...acc, [key]: value }), <T>{});
+}
```
Did Object.fromEntries<T> not work? I don't believe you need to write this.

hannahle replied:
No it didn't really work. Object.fromEntries < T > casts the value to whatever key to T.

kevinlul replied:
I see, then you do need to cast the return of that native function then, I misread the type signature.

kevinlul commented on server/src/resolvers/getVariantsResolver/utils/resolveChromosome.ts:
```diff
@@ -0,0 +1,14 @@
+const resolveChromosome = (position: string) => {
+  const chromosome = position.split(':')[0];
+  const coordinates = position.split(':')[1];
+  const end = coordinates.split('-')[0];
+  const start = coordinates.split('-')[1];
```
If it is actually this simple (that you can just split on these tokens assuming exactly one of them is in `position`), then you can also use a destructuring assignment to not split the same string twice:
```suggestion
  const [chromosome, coordinates] = position.split(':');
  const [end, start] = coordinates.split('-');
```

hannahle committed:
simplify resolveChromosome.ts

Co-authored-by: Kevin Lu <6320810+kevinlul@users.noreply.github.com>

hannahle committed:
reverted to use of Object.fromEntries with type casting

hannahle committed:
Merge branch 'hannah/tabix-wrapper' of github.com:ccmbioinfo/ssmp into hannah/tabix-wrapper

hannahle committed:
removed unused import
